onnxruntime/core/providers/shared_library/provider_interfaces.h

```diff
@@ -0,0 +1,480 @@
+      if (!CalcMemSizeForArray(count_or_bytes, sizeof(typename std::conditional<std::is_void<T>::value, void*, T>::type),
+                               &alloc_size)) return nullptr;
+    }
+    return Provider_IAllocatorUniquePtr<T>{
+        static_cast<T*>(allocator->Alloc(alloc_size)),  // allocate
+        [=](T* ptr) { allocator->Free(ptr); }};         // capture IAllocator so it's always valid, and use as deleter
+  }
+
+  void operator=(const Provider_IAllocator& v) = delete;
+};
+
+template <size_t alignment>
+bool Provider_IAllocator::CalcMemSizeForArrayWithAlignment(size_t nmemb, size_t size, size_t* out) noexcept {
+  static constexpr size_t max_allowed = (static_cast<size_t>(1) << (static_cast<size_t>(std::numeric_limits<size_t>::digits >> 1))) - alignment;
+  static constexpr size_t max_size = std::numeric_limits<size_t>::max() - alignment;
+  static constexpr size_t alignment_mask = alignment - 1;
+  //Indeed, we only need to check if max_size / nmemb < size
+  //max_allowed is for avoiding unnecessary DIV.
+  if (nmemb >= max_allowed && max_size / nmemb < size) {
+    return false;
+  }
+  if (size >= max_allowed &&
+      nmemb > 0 && max_size / nmemb < size) {
+    return false;
+  }
+  if (alignment == 0)
+    *out = size * nmemb;
+  else
+    *out = (size * nmemb + alignment_mask) & ~static_cast<size_t>(alignment_mask);
+  return true;
+}
+
+struct Provider_IDeviceAllocator : Provider_IAllocator {
+  virtual bool AllowsArena() const = 0;
+};
+
+using Provider_AllocatorPtr = std::shared_ptr<Provider_IAllocator>;
+using Provider_DeviceAllocatorFactory = std::function<std::unique_ptr<Provider_IDeviceAllocator>(int)>;
+
+struct Provider_DeviceAllocatorRegistrationInfo {
+  OrtMemType mem_type;
+  Provider_DeviceAllocatorFactory factory;
+  size_t max_mem;
+};
+
+class TensorShape;
+
+struct Provider_Tensor {
+  virtual float* MutableData_float() = 0;
+  virtual const float* Data_float() const = 0;
+
+  template <typename T>
+  T* MutableData();
+
+  template <typename T>
+  const T* Data() const;
+
+  virtual const TensorShape& Shape() const = 0;
+};
+
+template <>
+inline float* Provider_Tensor::MutableData<float>() { return MutableData_float(); }
+
+template <>
+inline const float* Provider_Tensor::Data<float>() const { return Data_float(); }
+
+struct Provider_OpKernelInfo {
+  virtual Status GetAttr(const std::string& name, int64_t* value) const = 0;
+  virtual Status GetAttr(const std::string& name, float* value) const = 0;
+
+  template <typename T>
+  Status GetAttr(const std::string& name, T* value) const;
+};
+
+template <>
+inline Status Provider_OpKernelInfo::GetAttr<int64_t>(const std::string& name, int64_t* value) const {
+  return GetAttr(name, value);
+}
+
+template <>
+inline Status Provider_OpKernelInfo::GetAttr<float>(const std::string& name, float* value) const {
+  return GetAttr(name, value);
+}
+
+struct Provider_OpKernelContext {
```

Comment:
Looks like we've 2 separate abstractions that wrap OpKernelContext for usage outside ORT - OrtKernelContext and this one. Does it make sense to unify them?

Reply:
Ideally, yes! One is for the C API and this one is internally using all of the C++ stuff directly. So one is ABI and this one isn't. But ideally yes.